pkg/core/native/notary.go

```diff
@@ -0,0 +1,358 @@
+	if err != nil {
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", addr.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+
+// withdraw sends all deposited GAS for "from" address to "to" address.
+func (n *Notary) withdraw(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	ok, err := runtime.CheckHashedWitness(ic, from)
+	if err != nil {
+		panic(fmt.Errorf("failed to check witness for %s: %w", from.StringBE(), err))
+	}
+	if !ok {
+		return stackitem.NewBool(false)
+	}
+	to := from
+	if !args[1].Equals(stackitem.Null{}) {
+		to = toUint160(args[1])
+	}
+	deposit := n.getDepositFor(ic.DAO, from)
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if ic.Chain.BlockHeight() < deposit.Till {
+		return stackitem.NewBool(false)
+	}
+	cs, err := ic.DAO.GetContractState(n.GAS.Hash)
+	if err != nil {
+		panic(fmt.Errorf("failed to get GAS contract state: %w", err))
+	}
+	transferArgs := []stackitem.Item{stackitem.NewByteArray(n.Hash.BytesBE()), stackitem.NewByteArray(to.BytesBE()), stackitem.NewBigInteger(deposit.Amount), stackitem.Null{}}
+	err = contract.CallExInternal(ic, cs, "transfer", transferArgs, smartcontract.All, vm.EnsureIsEmpty, func(ctx *vm.Context) { // we need EnsureIsEmpty because there's a callback popping result from the stack
+		isTransferOk := ic.VM.Estack().Pop().Bool()
+		if !isTransferOk {
+			panic("failed to transfer GAS from Notary account")
+		}
+	})
+	if err != nil {
+		panic(fmt.Errorf("failed to transfer GAS from Notary account: %w", err))
+	}
+	if err := n.removeDepositFor(ic.DAO, from); err != nil {
+		panic(fmt.Errorf("failed to remove withdrawn deposit for %s from the storage: %w", from.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+
+// balanceOf returns deposited GAS amount for specified address.
+func (n *Notary) balanceOf(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	acc := toUint160(args[0])
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		return stackitem.NewBigInteger(big.NewInt(0))
+	}
+	return stackitem.NewBigInteger(deposit.Amount)
+}
+
+// expirationOf Returns deposit lock height for specified address.
+func (n *Notary) expirationOf(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	acc := toUint160(args[0])
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		panic(fmt.Errorf("deposit for %s not found", acc.StringBE()))
+	}
+	return stackitem.Make(deposit.Till)
+}
+
+// verify checks whether the transaction was signed by one of the notaries.
+func (n *Notary) verify(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	sig, err := args[0].TryBytes()
+	if err != nil {
+		panic(fmt.Errorf("failed to get signature bytes: %w", err))
+	}
+	tx := ic.Tx
+	if len(tx.GetAttributes(transaction.NotaryAssistedT)) == 0 {
+		return stackitem.NewBool(false)
+	}
+	if tx.Sender() == n.Hash {
+		payer := tx.Signers[1].Account
```

Comment:
This will fault easily now `if len(tx.Signers) == 0`, because attributes are checked after witnesses.

Reply:
And maybe we better leave only witness presence check in attribute check and move the rest of this logic into `verify`.